### ISSUE TITLE
Highly-parallel deployment with Terragrunt Provider Caching fails with "501 Not Implemented" error

### ISSUE THREAD

**Issue by dgibbard-cisco** [gruntwork-io/terragrunt#3284]
## Describe the bug

With the new terragrunt provider cache enabled in a highly-parallel deployment (5 projects in parallel -- so, 5 independent calls to "terragrunt run-all init" on the same host for different directories), I seem to get the following error:
```
│ Error: Failed to install provider
│ 
│ Error while installing hashicorp/aws v5.35.0: could not query provider
│ registry for registry.terraform.io/hashicorp/aws: failed to retrieve
│ authentication checksums for provider: 501 Not Implemented returned from
│ 127.0.0.1:42633
```

Is this the result of some remote-side registry Rate Limiting, or a terragrunt bug? If the former; it would be good to get an idea on what sort of limits are in place, or recommended ```TERRAGRUNT_PARALLELISM``` settings to avoid it if possible...

Also occasionally saw network connectivity errors too.

## Steps To Reproduce

Just running "terragrunt run-all init" across multiple projects in parallel on a single (large) build host.
Initial run had what looked like just the one provider version trip up with this error from what I could tell; removed the build host and provisioned a new one in the same address space, and now it looks like all init provider download calls fail with the 501 error via the TG cache; so i'm suspecting a remote registry limit of some kind...

I currently have set ```TERRAGRUNT_PARALLELISM=50```

## Expected behavior

Either:
* No errors from terragrunt init
* Recommendations on TERRAGRUNT_PARALLELISM to avoid rate limiting from remote registries

## Nice to haves

- [ ] Terminal output
- [ ] Screenshots

## Versions

- Terragrunt version: 0.63.3
- OpenTofu/Terraform version: TF 1.3.9
- Environment details (Ubuntu 20.04, Windows 10, etc.): Ubuntu 20.04 // 22.04

**Comment by levkohimins** [comment]
Hey @dgibbard-cisco, this might be similar to issue #3234, since Terragrunt Provider Cache does not have a rate limiting feature.

**Comment by dalgibbard** [comment]
@levkohimins I wondered that too, but I only get this issue with the TG cache enabled; if I unset the envvar, TG runs fine...
Do you have any recommended debug actions we can try to get more info for where the 501 might be originating from?

**Comment by levkohimins** [comment]
Could you please provide a full log running with the environment variable `TERRAGRUNT_LOG_LEVEL=trace`?

**Comment by dgibbard-cisco** [comment]
Have done, but quite a bit of sensitive info I can't share in the full output at the mo - the section of interest though is pretty straightforward:
```
Initializing provider plugins...
- Reusing previous version of hashicorp/aws from the dependency lock file
time=2024-07-24T15:25:02+01:00 level=trace msg=Cache server received request status=200 uri=/v1/providers//registry.terraform.io/hashicorp/aws/versions
time=2024-07-24T15:25:02+01:00 level=trace msg=Cache server received request status=200 uri=/v1/providers//registry.terraform.io/hashicorp/aws/5.35.0/download/linux/amd64
time=2024-07-24T15:25:02+01:00 level=trace msg=Cache server received request status=501 uri=/downloads/releases.hashicorp.com/terraform-provider-aws/5.35.0/terraform-provider-aws_5.35.0_SHA256SUMS
╷
│ Error: Failed to install provider
│
│ Error while installing hashicorp/aws v5.35.0: could not query provider
│ registry for registry.terraform.io/hashicorp/aws: failed to retrieve
│ authentication checksums for provider: 501 Not Implemented returned from
│ 127.0.0.1:59713
╵
```

**Comment by dgibbard-cisco** [comment]
I also am using the following ```~./terraformrc``` file:
```
plugin_cache_dir = "$HOME/.cache/terragrunt/providers"
disable_checkpoint = true
provider_installation {
  filesystem_mirror {
    path = "/home/myuser/.cache/terragrunt/providers"
    include = [
      "internal.external.net/*/*"
    ]
  }
  direct {
    exclude = [
      "internal.example.net/*/*"
    ]
  }
}
```

**Comment by dgibbard-cisco** [comment]
Stripped this config down, and using the default terraform install dir as the filesystem mirror seems to have fixed it!

```
disable_checkpoint = true
provider_installation {
  filesystem_mirror {
    path = "/home/myuser/.terraform.d/plugins"
    include = [
      "internal.external.net/*/*"
    ]
  }
  direct {
    exclude = [
      "internal.example.net/*/*"
    ]
  }
}
```

**Comment by levkohimins** [comment]
1. Please try the new version v0.63.3
2. You need to specify in `--terragrunt-provider-cache-registry-names` all the provider registry names that you cache. This is most likely the reason why filesystem_mirror does not work as you expected, for example: 
```
--terragrunt-provider-cache-registry-names example.com --terragrunt-provider-cache-registry-names registry.opentofu.org --terragrunt-provider-cache-registry-names registry.terraform.io
```
